### DISCUSSION TITLE
Autosaving in game time rather than real time

### DISCUSSION THREAD

**Discussion by michalc** [OpenTTD/OpenTTD#12496] [Q&A]
OpenTTD 14.0 now does autosaving based on real time rather than game time, from https://github.com/OpenTTD/OpenTTD/pull/10655 I think. However, https://github.com/michalc/OpenTTDLab was using the autosave behaviour in game time to extract results reproducibly from OpenTDD, so it no longer works with 14.0 :-(

I'm wondering is there a way to restore autosaving based on game time? Maybe by a game script?

**Comment by glx22** [comment]
Main issue with using game time was it didn't save anything at all during pause, even when enabling "build while paused".
And with wallclock mode and freezed time the date doesn't move either.

**Reply by 2TallTyler** [reply]
We could use economy time and in wallclock mode describe it as "save at end of period". Still doesn't solve the pause/fast-forward issue and is somewhat a duplicate to real-time autosave in most cases besides this.

**Reply by michalc** [reply]
Thanks for your replies - it's sounding a bit like there isn't a game script way of saving every X in game time?

In terms of changes to OpenTTD, could both game and real time be supported? So there would be 9 different options for autosaving?

Off
Every 10 minutes (real time)
Every 30 minutes (real time)
Every 60 minutes (real time)
Every 120 minutes (real time)
Every month (game time)
Every 3 months (game time)
Every 6 months (game time)
Every 12 months (game time)

?

**Reply by michalc** [reply]
Oh! Or if it's not possible right now with game scripts... I wonder what it would take to make it possible? So the UI for players would be kept as is in 14.0, because it's sounding like this is the nicest when playing, but the behaviour I would like could be possible with a game script?

**Reply by michalc** [reply]
(I am happy to give things a go in terms of changes, but I might need guidance...)

**Reply by michalc** [reply]
I'm starting to wonder if this is possible with a gamescript right now... "all" it would need to do is issue the console save command every X gametime. Is this possible?

**Reply by michalc** [reply]
For the "every X gametime", say if want it the 1st of the month:
```squirrel
local done_on_date = null;
local current_date = null;
while(True) {
   current_date = GSDate.GetCurrentDate();
   if (current_date != done_on_date && GSDate.GetDayOfMonth(current_date) == 1) {
      /// Somehow save...
      done_on_date = current_date;
   }
   this.Sleep(1);
}
```
So need to figure out how to make it save...
(And maybe also get it optimized a bit? I wonder if it can predict how many ticks to sleep for...)

**Reply by michalc** [reply]
So I hope it's ok, but I made a PR that from a bit of testing, seems to work in terms of allowing game scripts to trigger saves: https://github.com/OpenTTD/OpenTTD/pull/12750

**Reply by michalc** [reply]
The game script way in https://github.com/OpenTTD/OpenTTD/pull/12750 was too open to be abused, so after discussion, going for a console script way, allowing scheduling of any console script: https://github.com/OpenTTD/OpenTTD/pull/12761
